pydicom/pixel_data_handlers/numpy_handler.py

```diff
@@ -101,7 +104,18 @@ def get_pixeldata(dicom_dataset):
 
     pixel_bytearray = dicom_dataset.PixelData
 
-    pixel_array = numpy.frombuffer(pixel_bytearray, dtype=numpy_dtype)
+    if dicom_dataset.BitsAllocated == 1:
+        # if single bits are used for binary representation, a uint8 array
+        # has to be converted to a binary-valued array (that is 8 times bigger)
+        try:
+            pixel_array = numpy.unpackbits(
+                numpy.frombuffer(pixel_bytearray, dtype='uint8'))
```

Comment:
Sorry for not catching this earlier, but this might not be the correct way to unpack the bits. In DICOM, the earlier bits fill the least significant bit first. See [part 5 section 8.1.1](http://dicom.nema.org/medical/dicom/current/output/chtml/part05/chapter_8.html#sect_8.1.1), and explanation in [Annex D](http://dicom.nema.org/medical/dicom/current/output/chtml/part05/chapter_D.html). Quoting from the latter:
> One way to visualize this packed encoding is to imagine encoding the Pixel Cells as a concatenated stream of bits from the least significant bit of the first Pixel Cell up through the most significant bit of the last Pixel Cell. Within this stream, the most significant bit of any Pixel Cell is followed by the least significant bit of the next Pixel Cell.

In a python snippet from [this comment](https://github.com/QIICR/dcmqi/issues/341#issue-308384664), I have (what I believe) is the right way to unpack the bits.

For what it's worth, the output of that snippet matches the output of the DCMTK dcm2pnm tool for the dataset attached with this comment (second of the 3 frames that are present in the attached multiframe SEG dataset):

```
........XXXX...........
.......XXXXXX........XX
....XXXXXXXXXX........X
X.XXXXXXXXXXXXX........
XXXXXXXXXXXXXXXX.......
.XXXXXXXXXXXXXXXXXX....
.XXXXXXXXXXXXXXXXXXX...
XXXXXXXXXXXX.XXXXXXXX..
XXXXXXXXXXXX..XXXXX.XX.
XXXXXXXXXXXXX..XXXX..XX
XXXXXXXXXXXXXX..XXX...X
XXXXXXXXXXXXXXX.XXX....
XXXXXXXXXXXXXXXXXXX....
XXXXXXXXXXXXXXXXXXX....
.XXXXXXXXXXXXXXXXXXX...
.XXXXXXXXXXX.XXXXX.XX..
..XXXXXXXXXX..XXXX..XX.
...XXXXXXXXXX..XX....XX
....XXXXXXXXXX........X
X.....XX.XXXXX.........
XX.....................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
```

DCMTK `dcm2pnm`:

![image](https://user-images.githubusercontent.com/313942/38339042-37363808-383a-11e8-9cf4-61a5d8a2b12d.png)

Output while using `numpy.unpackbits` for the same dataset:
```
........XXXX...........
.........XXXXXXXX......
XXXX......XXXXXXXX.....
.XXXXXXX...XXXXXXXX....
..XXXXXXXX..XXXXXXXX...
...XXXXXXXXXXXXXXXXXXX.
....XXXXXXXXXXXXXXXXXXX
XX....XXXXXXXXXXXXXXXXX
XXXX....XXXXXXXX.XXXXXX
XXXXXX...XXXXXXXX..XXXX
XXXXXXXX..XXXXXXXX...XX
XXXXXXXXXX.XXXXXXXX....
XXXXXXXXXXXXXXXXXXXXX..
..XXXXXXXXXXXXXXXXXXXX.
....XXXXXXXXXXXXXXXXXXX
X.....XXXXXXXXX.XXXXXXX
XX......XXXXXXXX..XXXXX
XXX......XXXXXXXX....XX
XXXX......XXXXXXXX.....
..XX........XXXXXXX....
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
.......................
```

[nonsquare_seg.dcm.zip](https://github.com/pydicom/pydicom/files/1878060/nonsquare_seg.dcm.zip)

For the sake of completeness, here is the python code from the issue referenced above that runs both approaches to unpack the bits:

```python
import pydicom, sys

d = pydicom.read_file(sys.argv[1])
if len(sys.argv)>2:
  frame = int(sys.argv[2])-1
  print("Dumping frame "+str(frame))
else:
  frame = None
print(d.Rows)
print(d.Columns)
print(d.NumberOfFrames)

totalPixels = int(d.Rows*d.Columns*d.NumberOfFrames/8)
if totalPixels%8:
  totalPixels = totalPixels + 1
totalPixels = totalPixels + (totalPixels % 2)
print("Total pixels expected: %i" % totalPixels)

print("Total pixels actual: %i" % len(d.PixelData))

if not frame is None:
  frames = [frame]
else:
  frames = range(d.NumberOfFrames)



import numpy as np
unpacked = np.unpackbits(np.frombuffer(d.PixelData,dtype=np.uint8))

print("With numpy unpackbits:")

for f in frames:
  for i in range(d.Rows):
    for j in range(d.Columns):
      pixelNumber = f*d.Rows*d.Columns+i*d.Columns+j
      if unpacked[pixelNumber]:
        sys.stdout.write("X")
      else:
        sys.stdout.write(".")
    print("")

print("\nWith manual unpacking:")
for f in frames:
  print("Frame %i" % f)
  for i in range(d.Rows):
    for j in range(d.Columns):
      pixelNumber = f*d.Rows*d.Columns+i*d.Columns+j
      byteNumber = int(pixelNumber/8)
      bitPosition = pixelNumber % 8
      if (d.PixelData[byteNumber] >> bitPosition) & 1:
        sys.stdout.write("X")
      else:
         sys.stdout.write(".")
    print("")
```

Reply:
I can look into contributing a patch to do this right later, but I wanted to mention it now so I don't forget, since I need to run now.